passes/cmds/splitfanout.cc

```diff
@@ -0,0 +1,244 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2012  Claire Xenia Wolf <claire@yosyshq.com>
+ *                2024  Akash Levy        <akash@silimate.com>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "kernel/yosys.h"
+#include "kernel/sigtools.h"
+#include "kernel/utils.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct SplitfanoutWorker
+{
+	Module *module;
+	SigMap sigmap;
+	dict<SigBit, tuple<IdString,IdString,int>> bit_drivers_db;
+	dict<SigBit, pool<tuple<IdString,IdString,int>>> bit_users_db;
```

Comment:
This construction is copied from `splitcells.cc`, it would be best to leave a comment that marks it as such for the sake of a future rework touching both instances of this pattern. I think some rework is possible to reduce the usage of these indices and make them more readable, but for the most part, if the runtime and memory consumption of the current implementation of `splitfanout` is sufficient for your use case then there isn't any immediate need for that.

Some of the possible improvements:
+ In buffer-normalized mode (see `bufnorm.cc`), `bit_drivers_db` wouldn't be needed
+ The construction of `toposort`, construction of the indices, and final cell duplication could be done with fewer passes over the cells. Though this would also make the pass harder to modify, read, maintain

Reply:
Thanks for noting these. We'll keep these in mind if performance issues arise in the future.

Reply:
I will add comments for the parts copied from `splitcells`